File: apps/web/src/api/playbooks.ts
import type {
  CreatePlaybookRequest,
  PlaybookListResponse,
  PlaybookVersion,
} from "@replaytutor/contracts";
import { validateContract } from "@replaytutor/contracts";
import { API_BASE_URL, ApiResponseError } from "./health";

async function checked<T>(
  response: Response,
  contract: Parameters<typeof validateContract>[0],
): Promise<T> {
  if (!response.ok) throw new ApiResponseError(`Playbook request failed: ${response.status}`, response.status);
  const value: unknown = await response.json();
  if (!validateContract<T>(contract, value)) throw new ApiResponseError(`Invalid ${contract} response`, 502);
  return value;
}

export async function fetchPlaybooks(): Promise<PlaybookListResponse> {
  return checked(await fetch(`${API_BASE_URL}/api/v1/playbooks`), "PlaybookListResponse");
}

export async function createPlaybook(payload: CreatePlaybookRequest): Promise<PlaybookVersion> {
  return checked(
    await fetch(`${API_BASE_URL}/api/v1/playbooks`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
    "PlaybookVersion",
  );
}

File: apps/api/replaytutor/routes/playbooks.py
from fastapi import APIRouter, Request

from replaytutor.config import Settings
from replaytutor.contracts import (
    CreatePlaybookRequest,
    PlaybookListResponse,
    PlaybookVersion,
)
from replaytutor.modules.playbook import PlaybookService

router = APIRouter(prefix="/api/v1/playbooks", tags=["playbooks"])


def service(request: Request) -> PlaybookService:
    settings: Settings = request.app.state.settings
    return PlaybookService(settings)


@router.get("", response_model=PlaybookListResponse)
def list_playbooks(request: Request) -> PlaybookListResponse:
    return service(request).list()


@router.post("", response_model=PlaybookVersion)
def create_playbook(
    request: Request,
    payload: CreatePlaybookRequest,
) -> PlaybookVersion:
    return service(request).create(payload)
